docs/contributing/999-rfcs/2024-03-24-snapshot-tests.md

```diff
@@ -0,0 +1,73 @@
+---
+
+# Cloud Snapshot Tests
+
+- **Author(s)**: @eladb
+- **Submission Date**: 2024-03-24
+- **Stage**: Proposal
+- **Stage Date**: 2024-03-24
+
+## Requirements
+
+We would like to be able to capture a snapshot of the synthesized output when compiling to cloud
+targets, and use it to protect against regressions in CI, without having to deploy to the cloud.
+
+For example, let's say I am creating a Wing library with a resource that supports both `sim` and
+`tf-aws`. I am writing a bunch of tests for it. 
+
+I can run these tests in the simulator (`wing test`) and can also run them on AWS through `wing test
+-t tf-aws`.
+
+As my code **or its dependencies** are updated, I'd like to continuously run all these tests to make
+sure my library is not broken. Running the tests against `sim` is easy (I can just run `wing test`
+in my CI build environment), but it's not realistic to run all these tests against the real cloud
+target, given it could take a very long time and could also be very expensive.
+
+In some scenarios (such as the Wing SDK tests), users will prefer to setup a continuous integration
+workflow and run tests in the cloud, but for the most part running all my tests on the cloud for
+every commit is too expensive and requires a complicated setup (i.e. an AWS account), so I'd like to
+use a more lightweight approach...
+
+> NOTE: This is not https://github.com/winglang/wing/issues/1214 (which is about using snapshots for
+> assertions).
+
+## Design
+
+We propose to add a feature to `wing test` which will capture a snapshot of the synthesized output
+of every `.test.w` file and store them in the repository.
+
+> This is a similar approach used successfully and scalably in the AWS CDK. See
+> [docs](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md) for details. The main
+> difference is that `cdk-integ` there was no way to add assertions to tests, so this is so much
+> cooler.
+
+When I run `wing test -t tf-aws foo.test.w` in my dev environment, if the test succeeds (e.g.
+deploys and all test cases pass), the test framework will create `foo.test.w.tfaws.snap` which will
+include a nicely formatted (markdown?) snapshot of all the synthesis output from this test.
+
+Now, when `wing test -t tf-aws foo.test.w` is executed in CI (`CI=1`), by default it won't actually
+go and deploy to the cloud. Instead, it will just synthesize the output and compare it to the
+committed `.snap` file. If the output is not the same, the test will fail with a nice diff
+indicating that there was an unexpected regression.
+
+The `--snapshots` or `-s` switch can be used to control behavior:
+
+ * `--snapshots=auto` - auto-detect based on CI flag (described above)
+ * `--snapshots=never` - disables snapshots altogether
+ * `--snapshots=dry` - will skip deployment and only update the snapshots.
+ * `--snapshots=wet` - forces a deployment even if `CI=1`.
+ * `--snapshots=assert` - only asserts that the snapshots have not changed
```

Comment:
maybe its just me but something feels off about `wing test -t tf-aws foo.test.w --snapshot=wet` since its the deployment aspect that is `wet` or `dry`.  Its essentially the "test" that is being performed in a dry or wet manner, not the snapshot.

I dont know If I have a better suggestion off the top of my head, but I almost feel like `wing test -t tf-aws foot.test.w --snapshot` should be the flag to force snapshot mode otherwise the deployment occurs.

Reply:
I see what you are saying... 

I played around with the idea of a boolean flag but since there are multiple modes  here, you need more than just a boolean...

So we need a way to model all of these modes somehow... Any suggestion?

Reply:
> So we need a way to model all of these modes somehow... Any suggestion?

hmm, not super sure. I mean the first few things that come to mind are either building a snapshot framework/library and control snapshots similar to how it works in jest.

Or maybe this is easier to model from a command perspective, `wing snapshot` 

Dunno need to think about it some

Reply:
Eh maybe I’m over analyzing on this datapoint I’m fine to try this and see